src/js/panels.js

```diff
@@ -198,16 +193,317 @@ https://github.com/fluid-project/infusion/raw/master/Infusion-LICENSE.txt
+            // The line below to add the scrolled distance of the parent container, which is "controlsDivScrollTop",
+            // rather than using button.offset().top only, is to fix an issue in Chrome and Safari that button.offset().top
+            // returns inconsistent value. The returned value sometimes has "controlsDivScrollTop" added, sometimes not.
+            // This line ensures consistent top values for the calculation to base upon.
+            currentButtonTop = currentButton.offset().top + controlsDivScrollTop,
+            closestPosition = gpii.firstDiscovery.panel.lang.findClosestNumber(currentButtonTop - that.lastMovedHeight, that.buttonTops),
+            heightToMove = currentButtonTop - closestPosition;
+
+        $(that.options.selectors.controlsDiv).animate({scrollTop: heightToMove + "px"}, 0);
+
+        that.lastMovedHeight = heightToMove;
+    };
+
+    gpii.firstDiscovery.panel.lang.getButtonTops = function (that) {
+        // setTimeout() is to work around the issue that position() in synchronous calls receives 0 for initial button positions
+        // when the panel is in the middle of rendering.
+        setTimeout(function () {
+            var buttons = that.locate("langRow"),
+                numOfLangPerPage = that.options.numOfLangPerPage;
+
+            // Keep track of the original positions of buttons on display
+            if (!that.buttonTops) {
+                that.buttonTops = [];
+                for (var i = 0; i < numOfLangPerPage; i++) {
+                    if (buttons[i]) {
+                        that.buttonTops[i] = $(buttons[i]).position().top;
+                    }
+                }
+                that.events.onButtonTopsReady.fire();
+            }
+        });
+    };
+
+    gpii.firstDiscovery.panel.lang.resetButtonTops = function (that, shownPanelId) {
+        var langPanelId = that.container.attr("id");
+        if (langPanelId === shownPanelId) {
+            that.buttonTops = undefined;
+            gpii.firstDiscovery.panel.lang.getButtonTops(that);
+        }
+    };
+
+    gpii.firstDiscovery.panel.lang.overrideDefaultScroll = function (that) {
+        that.locate("controlsDiv").scroll(function () {
+            gpii.firstDiscovery.panel.lang.scrollLangIntoView(that);
+        });
+    };
+
+    gpii.firstDiscovery.panel.lang.stopArrowBrowseOnEdgeButtons = function (button, keyCodes) {
+        $(button).keydown(function (e) {
+            if (keyCodes.indexOf(e.which) !== -1) {
+                e.preventDefault();
+                return false;
+            }
+        });
+    };
+
+    // When the focus is on the first language button, prevent that the press of up or left arrow keys to move to the last language button;
+    // when the focus is on the last language button, prevent that the press of down or right arrow keys to move to the first language button.
+    gpii.firstDiscovery.panel.lang.preventWrapWithArrowKeys = function (that) {
```

Comment:
There are some issues with using fluid.selectable() to control the arrow keys' movement among input radio buttons. One major issue is, once jQuery UI takes control of the focusing and movements, the selection change using arrow keys no longer triggers the radio button to be checked, as a result, doesn't trigger the language change. I believe this can be solved by adding select event option in fluid.selectable(), as well as for other issues. But just feel it's more clear and straight forward by using the current method of preventDefault() instead of adding another layer of a plugin and solve issues introduced by the plugin. What do you think?